src/protocol/sort/generate_permutation.rs

```diff
@@ -62,13 +69,26 @@ pub(super) async fn shuffle_and_reveal_permutation<
     )
     .await?;
 
-    // TODO: THIS IS WHERE VALIDATOR WILL BE CALLED!
+    let mut semihonest_context = None;
+    if let Some(malicious_validator) = malicious_validator {
+        semihonest_context = Some(
+            malicious_validator
+                .validate::<MaliciousReplicated<F>>(None)
```

Comment:
Rather than allow this to take an `Option`, we should provide a method that does nothing.  Though maybe that is not what is going on here.  We might be able to pass this the revealed permutation.  Nothing states that the Downgrade needs to do anything...

Reply:
Carrying that thought forward...

If we make `Downgrade` an async trait, then you might consider a version where your secret-shared permutations are "downgraded" by revealing them.  We could pass in the malicious context to the downgrade function, even, allowing you to invoke `reveal()` using that context.

That's a lot more complicated, but I think that it hides some of the complexity.  What do you think?

Reply:
This is a great suggestion! Let me try to code this and see if I can make it work

Reply:
I tried to make the Downgrade trait async and it propagated to validate, and ThisCodeIsAuthorizedToDowngradeFromMalicious. Still trying to resolve errors, diff can be found at https://github.com/richajaindce/raw-ipa/tree/async_malicious